design_notes/XDN03 - xlang Type System.md

```diff
@@ -15,3 +15,252 @@ status: draft
+events) but does not specify an implementation for those object members.
+
+Interfaces can be implemented by multiple different classes. Each class implementation of a given
+interface's object members is unrelated to any other class's implementation of the same interface.
+
+Interfaces may optionally inherit from a single other interface. All of the type members of the
+base interface are considered part of the derived interface as well. Any class implementing an
+interface that uses inheritance must implement all of the type members specified in all of the
+interfaces in the inheritance tree.
+
+> Note, WinRT interfaces do not support true inheritance. Rather, WinRT interfaces may optionally
+declare that they require one or more other interfaces to additionally be implemented on a class.
+This approach leads to inefficiencies at the ABI layer - both runtime overhead that stems from
+having to call QueryInterface as well as memory overhead that stems from types having multiple
+interface virtual memory tables. By limiting interfaces at the type system level to single
+inheritance, xlang can implement them at the ABI layer using a single composite virtual method
+table, avoiding the runtime and memory inefficiencies mentioned above.
+
+All xlang interfaces need an GUID identifier. This identifier has no relevance at the type system
+level, it is only used at the ABI level. An interface's GUID can either be explicitly provided or
+it can be implicitly generated from the interface's fully namespace qualified name using the
+algorithm defined in [RFC4122 section 4.3](https://tools.ietf.org/html/rfc4122#section-4.3).
+
+Interfaces are not versionable. Once they have been defined, they may never be changed.
+
+### Parameterized Interfaces
+
+xlang interfaces support type parameterization. Parameterized interfaces specify one or more type
+parameters that can be used as parameter types or return values of type members. Parameterized
+interface members may declare argument and/or return value types in terms of the type parameters
+instead of as concrete types.
+
+Like non-parameterized interfaces, parameterized interfaces require a GUID identifier. This GUID can
+be explicitly specified or implicitly generated from the type name.
+
+Parameterized interfaces can inherit from both parameterized and non-parameterized interfaces.
+When inheriting from a parameterized interface, the type parameters of the base interface must
+be specified, though they may be specified in terms of the derived interface's type parameters.
+
+A concrete interface is either a non-parameterized interface or a parameterized interface where all
+of the type parameters have been fully specified. Non-parameterized interfaces may inherit from
+concrete interfaces, but they may not inherit from a parameterized interface.
+
+#### Parameterized Interface Examples
+
+For example, xlang's foundation types includes a set of parameterized interfaces to represent
```

Comment:
Since we haven't dug into the collection space and how similar / different xlang is from WinRT, should this have a comment that the collection itnerfaces described here are preliminary / placeholders?

Reply:
I will clarify to indicate this document is not specifying the collection design. I recently added a new [stub XDN](https://github.com/Microsoft/xlang/blob/master/design_notes/XDN07%20-%20xlang%20Foundation%20Types.md#collections) to capture foundation type design (including collections)

Reply:
added link to XDN 07 here